core/src/test/scala/org/apache/spark/executor/ResourceDiscovererSuite.scala

```diff
@@ -0,0 +1,124 @@
+      Files.write("echo 0,1", file1, StandardCharsets.UTF_8)
+      JavaFiles.setPosixFilePermissions(file1.toPath(),
+        EnumSet.of(OWNER_READ, OWNER_EXECUTE, OWNER_WRITE))
+      sparkconf.set(EXECUTOR_GPU_DISCOVERY_SCRIPT, file1.getPath())
+      val resources = ResourceDiscoverer.findResources(sparkconf, false)
+      val gpuValue = resources.get(ResourceInformation.GPU)
+      assert(gpuValue.nonEmpty, "Should have a gpu entry")
+      assert(gpuValue.get.getCount() == 2, "Should have 2")
+      assert(gpuValue.get.getName() == ResourceInformation.GPU, "name should be gpu")
+      assert(gpuValue.get.getUnits() == "", "units should be empty")
+      assert(gpuValue.get.getAddresses().size == 2, "Should have 2 indexes")
+      assert(gpuValue.get.getAddresses().deep == Array("0", "1").deep, "should have 0,1 entries")
+    } finally {
+      JavaFiles.deleteIfExists(file1.toPath())
+    }
+  }
+
+  test("Resource discoverer multiple gpus driver") {
+    val sparkconf = new SparkConf
+
+    val file1 = File.createTempFile("test", "resourceDiscoverScript1")
+    try {
+      Files.write("echo 0,1", file1, StandardCharsets.UTF_8)
+      JavaFiles.setPosixFilePermissions(file1.toPath(),
+        EnumSet.of(OWNER_READ, OWNER_EXECUTE, OWNER_WRITE))
+      sparkconf.set(DRIVER_GPU_DISCOVERY_SCRIPT, file1.getPath())
+      sparkconf.set(EXECUTOR_GPU_DISCOVERY_SCRIPT, "boguspath")
+      // make sure it reads from correct config, here it should use driver
+      val resources = ResourceDiscoverer.findResources(sparkconf, true)
+      val gpuValue = resources.get(ResourceInformation.GPU)
+      assert(gpuValue.nonEmpty, "Should have a gpu entry")
+      assert(gpuValue.get.getCount() == 2, "Should have 2")
+      assert(gpuValue.get.getName() == ResourceInformation.GPU, "name should be gpu")
+      assert(gpuValue.get.getUnits() == "", "units should be empty")
+      assert(gpuValue.get.getAddresses().size == 2, "Should have 2 indexes")
+      assert(gpuValue.get.getAddresses().deep == Array("0", "1").deep, "should have 0,1 entries")
+    } finally {
+      JavaFiles.deleteIfExists(file1.toPath())
+    }
+  }
+
+
+  test("Resource discoverer script returns invalid number") {
+    val sparkconf = new SparkConf
+
+    val file1 = File.createTempFile("test", "resourceDiscoverScript1")
+    try {
+      Files.write("echo foo1", file1, StandardCharsets.UTF_8)
+      JavaFiles.setPosixFilePermissions(file1.toPath(),
+        EnumSet.of(OWNER_READ, OWNER_EXECUTE, OWNER_WRITE))
+      sparkconf.set(EXECUTOR_GPU_DISCOVERY_SCRIPT, file1.getPath())
+
+      val error = intercept[SparkException] {
+        ResourceDiscoverer.findResources(sparkconf, false)
+      }.getMessage()
+
+      assert(error.contains("The gpu discover script threw exception"))
+    } finally {
+      JavaFiles.deleteIfExists(file1.toPath())
+    }
+  }
+
+  test("Resource discoverer script doesn't exist") {
+    val sparkconf = new SparkConf
+
+    val file1 = new File("/tmp/bogus")
```

Comment:
Utils.withTempFile (forget its name)?

Reply:
there is a withTempDir, but don't see a withTempfile. In this particular case the file doesn't exist on purpose as it tests a non-existent file but could use in the above function.

Reply:
OK, that's fine, yeah. For other cases where you need a temp file I'd use Utils.withTempDir and make a file inside that dir